linkerd/dns/src/lib.rs

```diff
@@ -70,7 +70,7 @@ impl Resolver {
     ) -> Result<(Vec<net::SocketAddr>, time::Sleep), Error> {
         match self.resolve_srv(name).await {
             Ok(res) => Ok(res),
-            Err(e) if e.is::<InvalidSrv>() => {
+            Err(e) if e.is::<ResolveError>() || e.is::<InvalidSrv>() => {
```

Comment:
I believe the error returned by `resolve_srv` is currently going to _always_ be either a `ResolveError` (if the error was returned by `trust-dns-resolver`, or `InvalidSrv` (if the error was returned by `srv_to_socket_addr`).

The [`AsyncResolver::srv_lookup` function](https://docs.rs/trust-dns-resolver/latest/trust_dns_resolver/struct.AsyncResolver.html#method.srv_lookup) in `trust-dns-resolver` always returns `ResolveError`, and `srv_to_socket_addr` always returns `InvalidSrv`, and `resolve_srv` doesn't call any other fallible functions, so the error type is always one of these two errors.

That means we should be able to simplify this to just
```suggestion
            Err(e) => {
```
and remove the match arm on line 81 one that returns any other error from `resolve_srv`.

Reply:
Mmm, I think the explicit check is better. Otherwise, this is fragile if any of the other code changes. Otherwise, we should use an explicit error enum so that we can verify this statically.

Reply:
:woman_shrugging: hmm...I see your point. on the other hand, I'm not a huge fan of having the catchall match arm `Err(e) => Err(e)`, because it seems like it suggests that it's possible that we won't retry some error cases, which is misleading. 

maybe we should make that arm be `unreachable!()` or something instead, or remove the guards from the match arm and add
```rust
assert(e.is::<ResolveError>() || e.is::<InvalidSrv>())
```
in the body to make this explicit?

Reply:
honestly we could also just change the error types returned from `resolve_srv` and `srv_to_socket_addr` from a boxed error to `()` or something here, since we never actually care about any information contained in the error type...

Reply:
an error enum seems like the most complete solution here but I tried going down that path and it added a lot of boilerplate.  I'm hesitant to add `unreachable!()` or an assert because even though there aren't currently any other error types returned, if there were any, we would want them to return that error rather than panic.  In other words, that code path is currently unreachable, but it's not invalid.  With all this in mind, I'm inclined to leave it as-is.

Reply:
i'm not pulling for adding an error enum that we wouldn't use: we could just discard the errors entirely, which is _essentially_ what we're doing now, but with more steps...but i'm not going to block on this if no one else agrees with me.